src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -198,16 +182,18 @@ public ProjectCodeModel TryGetProjectCodeModel(ProjectId id)
         public EnvDTE.FileCodeModel GetOrCreateFileCodeModel(ProjectId id, string filePath)
             => GetProjectCodeModel(id).GetOrCreateFileCodeModel(filePath).Handle;
 
-        public void ScheduleDeferredCleanupTask(Action a)
-            => _deferredCleanupTasks.Add(_threadingContext.JoinableTaskFactory.StartOnIdle(a, VsTaskRunContext.UIThreadNormalPriority));
-
-        void IDisposable.Dispose()
+        public void ScheduleDeferredCleanupTask(Action<CancellationToken> a)
         {
-            // Stop any outstanding BG work we've queued up.
-            _cancellationTokenSource.Cancel();
-
-            // Now wait for all our cleanup tasks to finish before we return.
-            _deferredCleanupTasks.Join();
+            _ = _threadingContext.RunWithShutdownBlockAsync(async cancellationToken =>
+            {
+                await _threadingContext.JoinableTaskFactory.StartOnIdle(
```

Comment:
This double-layering wouldn't be required if you expose the `ShutdownBlockingFactory` from the threading context.

Reply:
Agreed, but covered by https://github.com/dotnet/roslyn/pull/45850#discussion_r456075192